crates/extensions/aws_mapper_ext/src/converter.rs

```diff
@@ -41,44 +46,71 @@ impl AwsConverter {
-                }
-
-                serde_json::to_string(&payload_json)?
-            } else {
-                return Ok(vec![]);
-            };
+            match self.mqtt_schema.entity_channel_of(&input.topic) {
+                Ok((source, channel)) => self.try_convert_te_topics(source, channel, input),
+                Err(_) => self.try_convert_tedge_topics(input),
+            }
+        }
+    }
 
+    fn try_convert_tedge_topics(
+        &mut self,
+        input: &MqttMessage,
+    ) -> Result<Vec<MqttMessage>, ConversionError> {
+        let default_timestamp = self.add_timestamp.then(|| self.clock.now());
+
+        // serialize with ThinEdgeJson for measurements, for alarms and events just add the timestamp
+        if input.topic.name.starts_with("tedge/events")
+            || input.topic.name.starts_with("tedge/alarms")
+            || input.topic.name.starts_with("tedge/health")
+        {
+            let mut payload_json: Map<String, Value> =
+                serde_json::from_slice(input.payload.as_bytes())?;
+
+            if let Some(timestamp) = default_timestamp {
+                let timestamp = timestamp
+                    .format(&time::format_description::well_known::Rfc3339)?
+                    .as_str()
+                    .into();
+                payload_json.entry("time").or_insert(timestamp);
+            }
+
+            let payload = serde_json::to_string(&payload_json)?;
             let topic_suffix = match input.topic.name.split_once('/') {
                 Some((_, topic_suffix)) => topic_suffix,
                 None => return Ok(vec![]),
             };
-
             let out_topic = Topic::new(&format!("aws/td/{topic_suffix}"))?;
 
             let output = MqttMessage::new(&out_topic, payload);
             self.size_threshold.validate(&output)?;
             Ok(vec![(output)])
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    fn try_convert_te_topics(
+        &mut self,
+        source: EntityTopicId,
+        channel: Channel,
+        input: &MqttMessage,
+    ) -> Result<Vec<MqttMessage>, ConversionError> {
+        match &channel {
+            Channel::Measurement {
+                measurement_type: m_type,
+            } => {
+                let default_timestamp = self.add_timestamp.then(|| self.clock.now());
+                // serialize with ThinEdgeJson for measurements, for alarms and events just add the timestamp
+                let mut serializer = ThinEdgeJsonSerializer::new_with_timestamp(default_timestamp);
+                tedge_api::parser::parse_str(input.payload_str()?, &mut serializer)?;
+                let source = normalize_name(source);
+                let out_topic = Topic::new(&format!("aws/td/{source}/m/{m_type}"))?;
+                let output = MqttMessage::new(&out_topic, serializer.into_string()?);
+                self.size_threshold.validate(&output)?;
+                Ok(vec![(output)])
```

Comment:
This is over complicated and doing the same as the roughly the same as `try_convert_tedge_topics`. i.e. changing the topic and adding a timestamp if missing.

The payload can be updated using `payload_json.entry("time").or_insert(timestamp);`
The topic can be rewritten in a generic way as the suffix is keep unchanged.

Reply:
If we do that, we might have to go with ` serde_json::from_slice(input.payload.as_bytes())?` converting all the payload values into a `Map with key & value` rather than the current way of parsing the payload using the `thin-edge serializer`. This might emit different kinds of errors with the wrong inputs. So, the existing tests have to be updated for error cases, is this what you are suggesting?

Reply:
>  is this what you are suggesting?

Yes